### ISSUE TITLE
Disable test suites that contain no tests

### ISSUE THREAD

**Issue by brson** [tikv/tikv#4391]
I haven't looked closely at this, but I'm hopeful that some compile time can be saved by not building empty test suites and doc test suites.

I think cargo has some ability to control which targets are built for testing. I hope it can control which are doc-tested.

So the task here is to find all of the empty test suites run by `cargo test --all` and disable them in Cargo.toml. Then time `cargo test --all --no-run` before and after.

**Comment by jswh** [comment]
Do you mean to add `#[cfg(test)]` to those test functions ?

**Comment by brson** [comment]
@jswh no, it's not that there are test cases that are missing the attribute, but that there are test suites with no test cases at all. And compiling them costs time. There is though a PR open for this.

**Comment by jswh** [comment]
I did some search but only found the `cargo test --all --exclude <spec>`  command that can exclude specific test suites. 
The empty test suties could be found by the test output. For example，the `test_storage` test suite has no cases as the output says
```
Running /mnt/c/Codes/third/tikv/target/debug/deps/test_storage-436a8955ad0b3b77

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```
The list was found with [a little script](https://gist.github.com/jswh/fd2a53f3bac33491cee0e27e6db525be) as

- fuzz
- bfuzz_targets
- fuzzer_afl
- fuzzer_honggfuzz
- fuzzer_libfuzzer
- panic_hook
- test_raftstore
- test_storage

And when I try to run tests with `time cargo test --all --no-run --exclude fuzz fuzzer_afl fuzzer_honggfuzz fuzzer_libfuzzer panic_hook, est_raftstore test_storage` it show some error
```
error: Found argument 'fuzzer_honggfuzz' which wasn't expected, or isn't valid in this context
```
I am not sure if it is what the wanted and what to do next.
Newbee here.

**Comment by brson** [comment]
I did another test, and found that disabling unused test suites did nothing for compile time. It makes me wonder if there's a bug in cargo, since I turned off _29_ test suites.

Here's my raw notes:

```
# 6/29

counting "0 test" suites

cargo test --all -- --list 2>&1 | tee ../all-tests.txt

rg "\d tests, \d* benchmarks" ../all-tests.txt

rg "\d tests, \d* benchmarks" ../all-tests.txt | wc -l

42 test / bench suites

rg "^0 tests, 0 benchmarks" ../all-tests.txt| wc -l

29 suites with no tests or benches

70% of test suites are empty

posted time-commits.sh https://gist.github.com/brson/14642390bfc05edf1fa2212ae7d609b4

branch: https://github.com/brson/tikv/tree/notest

cargo test --all --no-run full before
    Finished dev [unoptimized] target(s) in 12m 04s
cargo test --all --no-run partial before
    Finished dev [unoptimized] target(s) in 1m 24s
cargo test --all --no-run full after
    Finished dev [unoptimized] target(s) in 12m 16s
cargo test --all --no-run partial after
    Finished dev [unoptimized] target(s) in 1m 24s

```

**Comment by brson** [comment]
It does seem like disabling the tests prevented cargo from building the test suites - there are lots of binaries I don't see afterward. I'm super surprised it didn't reduce build time.
